rustler_codegen/src/init.rs

```diff
@@ -1,63 +1,125 @@
+        }
+
+        let get = |key| options.get(key).map(Clone::clone);
+
+        let load = get("load");
+        let unload = get("unload");
+        let upgrade = get("upgrade");
+
+        Ok(InitMacroInput {
+            name,
+            funcs,
+            load,
+            upgrade,
+            unload,
+        })
     }
 }
 
-fn parse_expr_assigns(input: ParseStream) -> Vec<syn::ExprAssign> {
-    let mut vec = Vec::new();
+fn parse_options(input: ParseStream) -> HashMap<String, Expr> {
+    let mut res = HashMap::new();
 
     while <Token![,]>::parse(input).is_ok() {
-        match syn::ExprAssign::parse(input) {
-            Ok(expr) => vec.push(expr),
-            Err(err) => panic!("{} (i.e. `load = load`)", err),
+        if input.is_empty() {
+            break;
         }
-    }
-    vec
-}
 
-fn extract_option(args: Vec<syn::ExprAssign>, name: &str) -> TokenStream {
-    for syn::ExprAssign { left, right, .. } in args.into_iter() {
-        if let syn::Expr::Path(syn::ExprPath { path, .. }) = &*left {
-            if let Some(ident) = path.get_ident() {
-                if *ident == name {
-                    let value = *right;
-                    return quote!(Some(#value));
+        match ExprAssign::parse(input) {
+            Ok(ExprAssign { left, right, .. }) => {
+                if let Expr::Path(ExprPath { path, .. }) = &*left {
+                    if let Some(ident) = path.get_ident() {
+                        res.insert(ident.to_string(), *right.clone());
+                    }
                 }
             }
+            Err(err) => panic!("{} (i.e. `load = load`", err),
         }
     }
-
-    let none = Ident::new("None", Span::call_site());
-    quote!(#none)
+    res
 }
 
 impl From<InitMacroInput> for proc_macro2::TokenStream {
     fn from(input: InitMacroInput) -> Self {
         let name = input.name;
         let num_of_funcs = input.funcs.elems.len();
         let funcs = nif_funcs(input.funcs.elems);
-        let load = input.load;
+
+        let nif_load_def = if let Some(load) = input.load {
+            quote! {
+                unsafe extern "C" fn nif_load(
+                    env: rustler::codegen_runtime::NIF_ENV,
+                    _priv_data: *mut *mut rustler::codegen_runtime::c_void,
+                    load_info: rustler::codegen_runtime::NIF_TERM
+                ) -> rustler::codegen_runtime::c_int {
+                    rustler::codegen_runtime::handle_nif_init_call(#load, env, load_info)
+                }
+                Some(nif_load)
+            }
+        } else {
+            quote!(None)
+        };
+
+        let nif_upgrade_def = if let Some(upgrade) = input.upgrade {
+            quote! {
+                unsafe extern "C" fn nif_upgrade(
+                    env: rustler::codegen_runtime::NIF_ENV,
+                    _priv_data: *mut *mut rustler::codegen_runtime::c_void,
+                    _old_priv_data: *mut *mut rustler::codegen_runtime::c_void,
```

Comment:
It doesn't make a whole lot of sense to not pass priv data into the provided upgrade function. But... `load` doesn't either so maybe it's fine?